src/core/config/config_module/merge.rs

```diff
@@ -0,0 +1,362 @@
+    C::Entry: FederatedMergeIntersection,
+{
+    fn federated_merge_intersection(mut self, other: Self) -> Valid<Self, String> {
+        Valid::from_iter(other, |(name, other_field)| {
+        match self.remove(&name) {
+            Some(field) => FederatedMergeIntersection::federated_merge_intersection(field, other_field).map(|merged| Some((name.clone(), merged))),
+            None => {
+                if other_field.type_of().is_nullable() {
+                    Valid::succeed(None)
+                } else {
+                    Valid::fail("Input arg is marked as non_null on the right side, but is not present on the left side".to_string())
+                }
+            },
+        }
+        .trace(&name)
+        })
+        .fuse(Valid::from_iter(self, |(name, field)| {
+            if field.type_of().is_nullable() {
+                Valid::succeed(())
+            } else {
+                Valid::fail("Input arg is marked as non_null on the left side, but is not present on the right side".to_string()).trace(&name)
+            }
+        }))
+        .map(|(merged_fields, _)| {
+            merged_fields.into_iter().flatten().collect()
+        })
+    }
+}
+
+impl<C> FederatedMergeUnion for C
+where
+    C: FederatedMergeCollection,
+    C::Entry: FederatedMergeUnion,
+{
+    fn federated_merge_union(mut self, other: Self) -> Valid<Self, String> {
+        Valid::from_iter(other, |(name, other_field)| match self.remove(&name) {
+            Some(field) => field
+                .federated_merge_union(other_field)
+                .map(|merged| (name.clone(), merged))
+                .trace(&name),
+            None => Valid::succeed((name, other_field)),
+        })
+        .map(|merged_fields| {
+            let mut merged_fields: C = merged_fields.into_iter().collect();
+            merged_fields.extend(self);
+
+            merged_fields
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs;
+
+    use insta::assert_snapshot;
+    use tailcall_fixtures::configs::federation;
+
+    use super::*;
+    use crate::core::config::{Config, ConfigModule};
+    use crate::core::valid::Validator;
+
+    #[test]
+    fn test_federation_merge() -> anyhow::Result<()> {
+        let config = Config::from_sdl(&fs::read_to_string(federation::ROUTER)?).to_result()?;
+        let router = ConfigModule::from(config);
+
+        let config =
+            Config::from_sdl(&fs::read_to_string(federation::SUBGRAPH_USERS)?).to_result()?;
+        let subgraph_users = ConfigModule::from(config);
+
+        let config =
+            Config::from_sdl(&fs::read_to_string(federation::SUBGRAPH_POSTS)?).to_result()?;
+        let subgraph_posts = ConfigModule::from(config);
+
+        let merged = router;
+        let merged = merged.federated_merge(subgraph_users).to_result()?;
+        let merged = merged.federated_merge(subgraph_posts).to_result()?;
+
+        assert_snapshot!(merged.to_sdl());
+
+        Ok(())
+    }
+}
```

Comment:
These are good E2E tests but it would be easier to comprehend if tests were added also for each usecase, for eg: one for union, for enums, and for types separately.